### DISCUSSION TITLE
request.session in HTTP Middleware

### DISCUSSION THREAD

**Discussion by sagar-arya** [fastapi/fastapi#6358] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI
from starlette.middleware.sessions import SessionMiddleware
app = FastAPI()
app.add_middleware(SessionMiddleware, secret_key="SECRET_KEY")

@app.middleware("http")
async def validate_user(request: Request, call_next):
    end_point = request.url.path
    request.session["name"] = "some random value"
    response = await call_next(request)
    return response
```

### Description

This is a basic example of my app.
I am trying to add values into the session if the user is valid.
For now, I am trying to set 
```
request.session["name"] = "some random value"
```
In the process, I get SessionMiddleware must be installed to access request.session
```
AssertionError: SessionMiddleware must be installed to access request.session
```
Is there any other way for solving it?
I am able to set it using
```
request.state.name="some random value"
```
I am not sure if this is a valid way to do it or not.
Can someone please help me with this?

### Operating System

macOS

### FastAPI Version

0.74.1

### Python Version

 3.10.2

**Comment by wmcgee3** [comment]
I was able to find this in the [Starlette docs](<https://www.starlette.io/middleware/#sessionmiddleware>): `Session information is readable but not modifiable.`
:cry: 

It looks like this package might provide the functionality you're looking for, though!
[FastAPI Sessions](<https://jordanisaacs.github.io/fastapi-sessions/>)

**Comment by akhilnarang** [comment]
Perhaps you could use [this](https://www.starlette.io/requests/#other-state)

**Comment by sagar-arya** [comment, marked as answer]
> I was able to find this in the [Starlette docs](https://www.starlette.io/middleware/#sessionmiddleware): `Session information is readable but not modifiable.` 😢
> 
> It looks like this package might provide the functionality you're looking for, though! [FastAPI Sessions](https://jordanisaacs.github.io/fastapi-sessions/)

@wmcgee3 tried all and going with request.state for now :)
closing the issue

**Comment by janczawadzki** [comment]
This may help: [order of middleware](https://github.com/encode/starlette/issues/479)

**Comment by elongstreet88** [comment]
I am still receiving this even just reading the session variable:

```python
@app.middleware("http")
async def validate_user(request: Request, call_next):
    end_point = request.url.path
    print(request.session) # <-- AssertionError: SessionMiddleware must be installed to access request.session
    response = await call_next(request)
    return response
```

How were you able to end up reading this?

**Comment by janczawadzki** [comment]
You need to make sure your middleware is attached to the app FIRST.  They are executed in reverse order - what is happening is that session hasn't run yet, so there is no request.session attribute (yet).  Get the middleware in the correct order, and you will always run AFTER the session middleware, and the problem goes away.

**Comment by elongstreet88** [comment]
@janczawadzki
Thanks for chiming in, as far as i can tell...i'm only using the session middleware and my minimal config looks like this:

```python
from fastapi import FastAPI, Request
from fastapi.responses import RedirectResponse, PlainTextResponse
from starlette.middleware.sessions import SessionMiddleware

app = FastAPI()
app.add_middleware(SessionMiddleware, secret_key="some-random-string")

@app.middleware("http")
async def validate_user(request: Request, call_next):
    print(request.session) # <--- Error: 'AssertionError: SessionMiddleware must be installed to access request.session'
    response = await call_next(request)
    return response

@app.get("/set/{item}")
async def set(request: Request, item:str):
    request.session["item"] = item
    return request.session["item"]

@app.get("/get")
async def get(request: Request) -> PlainTextResponse:
    return request.session.get("item", None)
```

Am i doing something obviously wrong here?
Thanks

**Comment by AKrumov** [comment]
> @janczawadzki Thanks for chiming in, as far as i can tell...i'm only using the session middleware and my minimal config looks like this:
> 
> ```python
> from fastapi import FastAPI, Request
> from fastapi.responses import RedirectResponse, PlainTextResponse
> from starlette.middleware.sessions import SessionMiddleware
> 
> app = FastAPI()
> app.add_middleware(SessionMiddleware, secret_key="some-random-string")
> 
> @app.middleware("http")
> async def validate_user(request: Request, call_next):
>     print(request.session) # <--- Error: 'AssertionError: SessionMiddleware must be installed to access request.session'
>     response = await call_next(request)
>     return response
> 
> @app.get("/set/{item}")
> async def set(request: Request, item:str):
>     request.session["item"] = item
>     return request.session["item"]
> 
> @app.get("/get")
> async def get(request: Request) -> PlainTextResponse:
>     return request.session.get("item", None)
> ```
> 
> Am i doing something obviously wrong here? Thanks

In order to fix this please move: `app.add_middleware(SessionMiddleware, secret_key="some-random-string")` on the last line. This will fix your problem :)

**Reply by felipequentino** [reply]
Thank you so much. GOAT
EVERY LLM told me the OPPOSITE.

**Comment by elongstreet88** [comment]
HA! That worked but is less than intuitive.... :)
Thanks.

**Comment by HarrisonFok** [comment]
@sagar-arya can you post your working code with fastapi-session?

**Comment by janczawadzki** [comment]
```
from starlette.middleware.base import BaseHTTPMiddleware
from starlette.responses import PlainTextResponse

class AuthMiddleware(BaseHTTPMiddleware):

    async def dispatch(self, request, call_next):

        request.state.foo = "bar"
        response = await call_next(request)
        return response
```

Now initialize the app: 

```
from fastapi import FastAPI
import middleware # the stuff above
app = FastAPI()
app.add_middleware(middleware.AuthMiddleware)
```


In your method calls...


```
@app.get('/foo')
async def foo(request:Request):
   logging.info(f"Foo {request.status.foo}")

```

**Reply by janczawadzki** [reply]
We don't use the session middleware, wrote our own - the middleware checks a session header and injects various things into the request.state

**Reply by dineshbvadhia** [reply]
@janczawadzki I want to store the client request during a user session. I know this can be achieved with an authentication framework. Does your code capture the users request?

**Comment by msingh0101** [comment]
Ran into this issue and even Claude.AI couldnt help solve it.  The note above about making sure SessionMiddleware is added last resolved my issues for other handlers that need to use session.   This begs the question, given it was so difficult to find a solution - is using Sessions a recommended solution for FastAPI apps  or is something like request.state what most use instead
